enhancements/node-tuning/mixed-cpu-node-plugin.md

```diff
@@ -0,0 +1,281 @@
+so having some latency is bearable.
+A way of mitigating that is checks whether it's possible to use workload partitioning
+to ensure the platform housekeeping processes don't run on the shared cpus.
+(This statement needs to be verified)
+
+* There is no risk that workload’s tasks run on dedicated OCP’s housekeeping reserved cpus,
+because only the shared cpus exposed via cgroups to the workload’s process.
+In other words, platform's housekeeping processes can run on shared cpus dedicated to workload’s light-weight tasks,
+but not the other way around.
+
+* mixed-cpus-node-plugin is NRI-based plugin.
+Bugs in the plugin or NRI framework might affect the containers at various lifecycle stages, e.g.,
+CreateContainer, UpdateContainer, etc.
+For mitigating such cases, it is possible to deactivate the node-plugin, and in severe cases it is possible to disable
+NRI completely at runtime-level.
+
+* From a performance perspective, the runtime communicates with the node-plugin via gRPC calls.
+Therefore, for every container creation, there's another gRPC call between the runtime and the node-plugin.
+We can't avoid this call even for containers that don't ask for shared CPUs, since runtime doesn't have any filtering
+mechanism, it sends all the containers through the NRI framework to node-plugin, and the filtering is done at the
+plugin side.
+
+### Drawbacks
+
+The way of how user should ask for shared CPUs in the pod spec is through device plugin.
+While this approach fits in the model of how the workload requires resources,
+better integration would require more invasive changes which are out of scope now.
+    
+## Design Details
+N/A
+
+### Test Plan
+
+#### Unit testing
+We will have unit-testing for the mixed-cpu-node-plugin to guarantee the basic functionality.
+
+#### E2E testing on mixed-cpu-node-plugin
+The behavior of the mixed-cpu-node-plugin will be validated in isolation (without NTO) by E2E tests
+
+#### E2E testing on NTO
+Validating a deployment process on NTO, verifying PerformanceProfile configuration and checking 
+the functionality of mixed-cpu-node-plugin (running the e2e tests from mixed-cpu-node-plugin) 
+
+### Graduation Criteria
+N/A
+
+#### Dev Preview -> Tech Preview
+N/A
+
+#### Tech Preview -> GA
+N/A
+
+#### Removing a deprecated feature
+N/A
+
+### Upgrade / Downgrade Strategy
+N/A
+
+### Version Skew Strategy
+N/A
+
+### Operational Aspects of API Extensions
+N/A
+
+#### Failure Modes
+N/A
+
+#### Support Procedures
+N/A
+
+## Implementation History
+N/A
+
+## Alternatives
+* Suggest a change to Kubelet CPU manager to support a static shared-cpu pool (not dynamic as of today).
+A change such as this requires changes to Kubelet, scheduler and other supporting controllers as 
+eviction manager, HPA (horizontal pod autoscaling), etc.
+Considering the u/s velocity, current deadlines, and the number of open questions that have to be addressed,
+the plugin solution has a bigger chance to be completed on time.
```

Comment:
I'd suggest that "bigger chance to be completed on time" isn't the best criteria for choosing an alternative. Another consideration is what are going to be the implications for migrating to a proper solution if this short-term approach is implemented?

Reply:
Hey Bart thank you for taking the time and providing your feedback.

> I'd suggest that "bigger chance to be completed on time" isn't the best criteria for choosing an alternative.

1. I would not underestimate the u/s pace which can be extremely slow, especially for centralize changes in resources management and pod spec API.
2. Time to market is not the only obstacle - In order to reach consensus among u/s community we should present a broader set of use-cases and scenarios which is relevant to wider audience, which we don't have atm, but the specific use-case mentioned in the user-story section.

>Another consideration is what are going to be the implications for migrating to a proper solution if this short-term approach is implemented?

That depends on what the proper solution would be, but let's assume the proper solution is having a new API
field in pod spec + new pool on kubelet config.
So upgrade from short-term to proper solution would roughly look like:
1. The mixed-cpu-node-plugin can be shut-off easily by a simple change in the performance profile.
2.  The new pool would be configured in Kubelet-config via NTO.
3. All pod spec that requests `openshift.io/shared-cpu` would be convert (via API admission hook IIRC) to the new API filed.